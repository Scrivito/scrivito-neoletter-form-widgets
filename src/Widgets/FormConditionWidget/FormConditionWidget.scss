.scrivito-neoletter-form-widgets .condition-info {
  position: absolute;
  top: 6px;
  left: 50%;
  background: white;
  transform: translate(-50%);
  padding: 0 5px;
  margin: 0;
}
