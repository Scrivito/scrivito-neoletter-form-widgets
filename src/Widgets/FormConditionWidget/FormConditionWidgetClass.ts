import * as Scrivito from "scrivito";

export const FormConditionWidget = Scrivito.provideWidgetClass(
  "FormConditionWidget",
  {
    onlyInside: "FormConditionalContainerWidget",
    attributes: {
      title: "string",
      content: "widgetlist"
    }
  }
);
