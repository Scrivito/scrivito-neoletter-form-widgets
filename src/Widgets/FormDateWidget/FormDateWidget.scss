.scrivito-neoletter-form-widgets .form-date {
  .date-title {
    p {
      display: inline;
    }
  }
}
