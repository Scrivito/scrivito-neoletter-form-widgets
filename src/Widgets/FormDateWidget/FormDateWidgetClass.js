import * as Scrivito from "scrivito";

export const FormDateWidget = Scrivito.provideWidgetClass("FormDateWidget", {
  attributes: {
    title: "string",
    helpText: "html",
    required: "boolean",
    customFieldName: "string",
    dateType: [
      "enum",
      {
        values: ["date", "datetime-local"],
      },
    ],
  },
});
