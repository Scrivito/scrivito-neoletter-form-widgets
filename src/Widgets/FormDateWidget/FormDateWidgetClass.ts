import * as Scrivito from "scrivito";

export const FormDateWidget = Scrivito.provideWidgetClass("FormDateWidget", {
  attributes: {
    title: "html",
    helpText: "html",
    required: "boolean",
    customFieldName: "string",
    dateType: [
      "enum",
      {
        values: ["date", "datetime-local"]
      }
    ]
  }
});
