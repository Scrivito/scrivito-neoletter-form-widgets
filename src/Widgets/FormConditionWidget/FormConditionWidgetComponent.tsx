import * as React from "react";
import * as Scrivito from "scrivito";
import { InPlaceEditingPlaceholder } from "../../Components/InPlaceEditingPlaceholder";
import { FormConditionWidget } from "./FormConditionWidgetClass";
import "./FormConditionWidget.scss";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
Scrivito.provideComponent(FormConditionWidget, ({ widget, getData }: any) => {
  const data = getData ? getData(widget.id()) : { isActive: false };

  if (!data.isActive && !Scrivito.isInPlaceEditingActive()) {
    return null;
  }

  if (!widget.get("content").length) {
    return (
      <InPlaceEditingPlaceholder center>
        Select widgets in the widget properties.
      </InPlaceEditingPlaceholder>
    );
  }

  return (
    <>
      {Scrivito.isInPlaceEditingActive() && (
        <span className="condition-info">
          {"Condition: " + widget.get("title")}
        </span>
      )}
      {widget.get("content") && widget.get("content").length > 0 && (
        <Scrivito.ContentTag
          content={widget}
          attribute="content"
          className={
            Scrivito.isInPlaceEditingActive() ? "conditional-content" : ""
          }
        />
      )}
    </>
  );
});
