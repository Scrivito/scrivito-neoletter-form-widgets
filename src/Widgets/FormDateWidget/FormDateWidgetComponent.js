import * as React from "react";
import * as Scrivito from "scrivito";
import { getFieldName } from "../FormContainerWidget/utils/getFieldName";
import { HelpText } from "../FormContainerWidget/components/HelpTextComponent";
import { Mandatory } from "../FormContainerWidget/components/MandatoryComponent";
import "./FormDateWidget.scss";

Scrivito.provideComponent("FormDateWidget", ({ widget }) => {
  const [value, setValue] = React.useState("");
  const onChangeValue = (e) => {
    const d = new Date(e.target.value);
    setValue(d.toISOString());
  };
  return (
    <div className="form-date mb-3">
      <div className="date-title">
        <span className="text-super">{widget.get("title")}</span>
        {widget.get("required") && <Mandatory />}
        {widget.get("helpText") && <HelpText widget={widget} />}
      </div>

      <input
        onChange={onChangeValue}
        className="datepicker"
        type={widget.get("dateType")}
        required={widget.get("required")}
      ></input>
      <input type="hidden" name={getFieldName(widget)} value={value}></input>
    </div>
  );
});
