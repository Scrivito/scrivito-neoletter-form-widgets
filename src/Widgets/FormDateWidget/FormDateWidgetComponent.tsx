import * as React from "react";
import * as Scrivito from "scrivito";
import { getFieldName } from "../FormStepContainerWidget/utils/getFieldName";
import { HelpText } from "../FormStepContainerWidget/components/HelpTextComponent";
import { Mandatory } from "../FormStepContainerWidget/components/MandatoryComponent";
import { FormDateWidget } from "./FormDateWidgetClass";
import "./FormDateWidget.scss";

Scrivito.provideComponent(FormDateWidget, ({ widget }) => {
  const [value, setValue] = React.useState("");
  const onChangeValue = (e: React.BaseSyntheticEvent) => {
    const d = new Date(e.target.value);
    setValue(d.toISOString());
  };
  return (
    <div className="form-date mb-3">
      <div className="date-title">
        <span className="text-super">{widget.get("title")}</span>
        {widget.get("required") && <Mandatory />}
        {widget.get("helpText") && <HelpText widget={widget} />}
      </div>

      <input
        onChange={onChangeValue}
        className="datepicker"
        type={widget.get("dateType")!}
        required={widget.get("required")}></input>
      <input
        type="hidden"
        className="show-in-review"
        name={getFieldName(widget)}
        value={value}></input>
    </div>
  );
});
