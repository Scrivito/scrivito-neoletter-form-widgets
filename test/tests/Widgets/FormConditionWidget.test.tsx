import * as Scrivito from "scrivito";
import { FormConditionWidget } from "../../../src/Widgets/FormConditionWidget/FormConditionWidgetClass";
import PageRenderer from "../../helpers/pageRenderer";
import { FormDateWidget } from "../../../src/Widgets/FormDateWidget/FormDateWidgetClass";
import "../../../src/Widgets/FormConditionWidget/FormConditionWidgetComponent";
import "../../../src/Widgets/FormDateWidget/FormDateWidgetComponent";

Scrivito.configure({ tenant: "inMemory" });
// mocking getData seems to be not possible, therefore isActive will always be false!
const pageRenderer = new PageRenderer();
const widgetProps = {
  title: "select me",
  content: []
};

describe("FormConditionWidget", () => {
  it("should not show content if not if isInPlaceEditingActive is false", () => {
    jest.spyOn(Scrivito, "isInPlaceEditingActive").mockReturnValue(false);
    pageRenderer.render({
      body: [new FormConditionWidget(widgetProps)]
    });

    const content = document.querySelector(".conditional-content");
    expect(content).not.toBeInTheDocument();
  });

  it("should  show condition-info & content if isInPlaceEditingActive is true", () => {
    jest.spyOn(Scrivito, "isInPlaceEditingActive").mockReturnValue(true);
    const dummyContent = [
      new FormDateWidget({
        customFieldName: "custom_date",
        title: "date",
        dateType: "date"
      })
    ];
    pageRenderer.render({
      body: [new FormConditionWidget({ ...widgetProps, content: dummyContent })]
    });

    const content = document.querySelector(".conditional-content");
    const condtionInfo = document.querySelector(".condition-info");
    expect(content).toBeInTheDocument();
    expect(condtionInfo).toBeInTheDocument();
    expect(condtionInfo).toHaveTextContent("Condition: " + widgetProps.title);
  });

  it("renders correctly", () => {
    const dummyContent = [
      new FormDateWidget({
        customFieldName: "custom_date",
        title: "date",
        dateType: "date"
      })
    ];
    const tree = pageRenderer.getAsJSON({
      body: [new FormConditionWidget({ ...widgetProps, content: dummyContent })]
    });
    expect(tree).toMatchSnapshot();
  });
});
