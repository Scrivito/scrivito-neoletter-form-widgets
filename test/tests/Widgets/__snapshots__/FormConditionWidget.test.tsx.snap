// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormConditionWidget renders correctly 1`] = `
<div>
  <div>
    <span
      className="condition-info"
    >
      Condition: select me
    </span>
    <div
      className="conditional-content"
    >
      <div>
        <div
          className="form-date mb-3"
        >
          <div
            className="date-title"
          >
            <span
              dangerouslySetInnerHTML={
                {
                  "__html": "date",
                }
              }
              onClick={[Function]}
            />
          </div>
          <input
            className="datepicker"
            onChange={[Function]}
            required={false}
            type="date"
          />
          <input
            className="show-in-review"
            name="custom_date"
            type="hidden"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
